### ISSUE TITLE
[clang-17 regression] clang from trunk crashes inside clang::TemplateArgument::pack_size()

### ISSUE THREAD

**Issue by loskutov** [llvm/llvm-project#63782]
The following piece of code leads to ICE when compiled with clang from trunk:

```cpp
template<bool... Vals>
constexpr bool All = (Vals && ...);

template<bool... Bs>
class Class {
    template<typename>
    requires All<Bs...>
    void Foo();
};

template<bool... Bs>
template<typename>
requires All<Bs...>
void Class<Bs...>::Foo() {
};
```

Compiler Explorer with full stacktrace: https://godbolt.org/z/n6bro9Tcd

**Comment by shafik** [comment]
@Fznamznon any chance your recent fix could be involved here?

**Comment by Fznamznon** [comment]
> @Fznamznon any chance your recent fix could be involved here?

I don't think so. I'm not seeing any `MemberExpr`s created (the patch tackles only them), also reverting the patch locally and seeing the same assertion failure seems pretty assuring.

**Comment by loskutov** [comment]
According to git-bisect, 6db007a0654ed7a6ed5c3aa3b61a937c19a6bc6b authored by @alexander-shaposhnikov is the first bad commit.

**Comment by michael-jabbour-sonarsource** [comment]
Another crashing example that results in a similar stacktrace.

```cpp
using size_t = decltype(sizeof(int));

template<typename... Ts> struct S {
  template<size_t N>
  requires(N <= sizeof...(Ts))
  void f(int);
};

template<typename... Ts>
template<size_t N>
requires(N <= sizeof...(Ts))
void S<Ts...>::f(int) {}

```
This also bisects to https://github.com/llvm/llvm-project/commit/6db007a0654ed7a6ed5c3aa3b61a937c19a6bc6b. Godbolt link: https://godbolt.org/z/s9YvEG5rn.

**Comment by shafik** [comment]
CC @alexander-shaposhnikov @erichkeane

**Comment by erichkeane** [comment]
I really hope @alexander-shaposhnikov can take a look at this, I've been away from this for long enough, and I likely wont have a chance to look at it 'soon'.

**Comment by alexander-shaposhnikov** [comment]
Hi, I'll try to look into this in the coming weeks, but I'm really busy with completely unrelated things at the moment / can't promise.

**Comment by erichkeane** [comment]
I spent a little time that I had to spare in the debugger with this one, but I'm a touch lost.  The function in question is: https://clang.llvm.org/doxygen/classclang_1_1Sema.html#a0b0f6c317ec8c9f08dcb406114c43644

The issue is at line 732: https://clang.llvm.org/doxygen/SemaTemplateVariadic_8cpp_source.html#l00732

In @michael-jabbour-sonarsource 's example, we're assuming the argument (the only one, at 0,0) is a pack, it is NOT.  It is a `PackExpansionType` (`Ts...`), which SEEMS right, but I have very little idea what is going on in this function, or why a `PackExpansionType` is not sufficiently 'pack like' here to work?  I'm hopeful @alexander-shaposhnikov has some luck on this, but my time is pretty limited unfortunately right now as well.

**Comment by erichkeane** [comment]
Further digging shows that just handling the `PackExpansionType` isn't sufficient.  I suspect Alexander's patch ended up putting the pack into the MLTAL in a way that it shouldn't have, but I don't see how/why.

**Comment by erichkeane** [comment]
The issue is here... SOMEWHERE: https://github.com/llvm/llvm-project/commit/6db007a0654ed7a6ed5c3aa3b61a937c19a6bc6b#diff-dae8060cd0876af2d1e1376f5da83d22a01fd1b49730cee2deaee9727e64e0e4R221

**Comment by alejandro-alvarez-sonarsource** [comment]
Hello.
I believe this was fixed by #80594. clang trunk with assertions does not crash anymore with the given example: https://godbolt.org/z/1TnqxbE1E

**Comment by shafik** [comment]
Looks fixed in trunk
